leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/impl/CaliforniumEndpointsManager.java

```diff
@@ -0,0 +1,198 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Sierra Wireless and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *     Sierra Wireless - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.leshan.client.californium.impl;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Collection;
+
+import org.eclipse.californium.core.CoapServer;
+import org.eclipse.californium.core.network.CoapEndpoint;
+import org.eclipse.californium.core.network.Endpoint;
+import org.eclipse.californium.core.network.config.NetworkConfig;
+import org.eclipse.californium.scandium.DTLSConnector;
+import org.eclipse.californium.scandium.config.DtlsConnectorConfig;
+import org.eclipse.californium.scandium.config.DtlsConnectorConfig.Builder;
+import org.eclipse.californium.scandium.dtls.pskstore.StaticPskStore;
+import org.eclipse.leshan.SecurityMode;
+import org.eclipse.leshan.client.servers.EndpointsManager;
+import org.eclipse.leshan.client.servers.Server;
+import org.eclipse.leshan.client.servers.ServerInfo;
+import org.eclipse.leshan.core.californium.EndpointFactory;
+import org.eclipse.leshan.core.request.Identity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CaliforniumEndpointsManager implements EndpointsManager {
```

Comment:
What is the responsibility of this class? What I see is that it will create a single endpoint for several servers, am I right?

Reply:
Creating/Starting/Stopping endpoints. This class depends on californium.
The registration engine does not depend on it and so uses the EndpointsManager abstraction.

This class should create 1 endpoint by server, but for now we don't support several dm server.
So it create only 1 endpoint for the bootstrap server in a bootstrapsession.
Once the bootstrap session is finished we create a new endpoint for the DM server (the only one we support for now)

I tried to put the brick to support multi-server (maybe it's a mistake and its too soon to do that)

Reply:
Then, my understanding is that we now use a single client per endpoint, right?

Reply:
This time that's me who is confused :p

At Leshan server side, nothing changes. (2 endpoints : 1 secured and 1 unsecured which allow to talk to any number of clients)

At Leshan client side, until now that was like as server side. But californium has limitation and does not allow several RPK or X509 certificate for the same coapEndpoint nor to change credential dynamically. So I tried this approach which recreates endpoint at bootstrap which will help to support X509 or RPK and  which could also help to support several dm server later. In that case the lwm2m client will have an endpoint per server.

But maybe https://github.com/eclipse/californium/tree/credentials_store will be a game-changer ...